File: apps/services/notifications/Notifications.Infrastructure/Services/WebhookIngestionService.cs
using System.Text.Json;
using Microsoft.Extensions.Logging;
using Notifications.Application.Interfaces;
using Notifications.Domain;
using Notifications.Infrastructure.Webhooks.Normalizers;
using Notifications.Infrastructure.Webhooks.Verifiers;
using LegalSynq.AuditClient;
using LegalSynq.AuditClient.DTOs;

namespace Notifications.Infrastructure.Services;

public class WebhookIngestionServiceImpl : IWebhookIngestionService
{
    private readonly IWebhookLogRepository _webhookLogRepo;
    private readonly INotificationEventRepository _eventRepo;
    private readonly INotificationAttemptRepository _attemptRepo;
    private readonly IDeliveryStatusService _deliveryStatusSvc;
    private readonly IRecipientContactHealthService _contactHealthSvc;
    private readonly IDeliveryIssueService _deliveryIssueSvc;
    private readonly IContactSuppressionRepository _suppressionRepo;
    private readonly ISmsPreferenceService _smsPreferenceSvc;
    private readonly IInboundSmsResolverService _inboundSmsResolver;
    private readonly SendGridVerifier _sendGridVerifier;
    private readonly TwilioVerifier _twilioVerifier;
    private readonly IAuditEventClient _auditClient;
    private readonly ILogger<WebhookIngestionServiceImpl> _logger;

    private static readonly HashSet<string> DeliveryFinalEvents = new() { "delivered", "bounced", "failed", "undeliverable", "rejected", "complained", "unsubscribed" };
    private static readonly HashSet<string> InboundDirections = new(StringComparer.OrdinalIgnoreCase) { "inbound", "inbound-api" };
    private static readonly Dictionary<string, string> SgSuppressionMap = new() { ["bounced"] = "bounce", ["complained"] = "complaint", ["unsubscribed"] = "unsubscribe" };
    private static readonly Dictionary<string, string> TwilioSuppressionMap = new() { ["bounced"] = "bounce", ["complained"] = "complaint", ["unsubscribed"] = "unsubscribe", ["carrier_rejected"] = "carrier_rejection" };

    public WebhookIngestionServiceImpl(
        IWebhookLogRepository webhookLogRepo, INotificationEventRepository eventRepo, INotificationAttemptRepository attemptRepo,
        IDeliveryStatusService deliveryStatusSvc, IRecipientContactHealthService contactHealthSvc, IDeliveryIssueService deliveryIssueSvc,
        IContactSuppressionRepository suppressionRepo, ISmsPreferenceService smsPreferenceSvc,
        IInboundSmsResolverService inboundSmsResolver,
        SendGridVerifier sendGridVerifier, TwilioVerifier twilioVerifier,
        IAuditEventClient auditClient, ILogger<WebhookIngestionServiceImpl> logger)
    {
        _webhookLogRepo = webhookLogRepo; _eventRepo = eventRepo; _attemptRepo = attemptRepo;
        _deliveryStatusSvc = deliveryStatusSvc; _contactHealthSvc = contactHealthSvc; _deliveryIssueSvc = deliveryIssueSvc;
        _suppressionRepo = suppressionRepo; _smsPreferenceSvc = smsPreferenceSvc;
        _inboundSmsResolver = inboundSmsResolver;
        _sendGridVerifier = sendGridVerifier; _twilioVerifier = twilioVerifier;
        _auditClient = auditClient; _logger = logger;
    }

    public async Task<WebhookResult> HandleSendGridAsync(string rawBody, Dictionary<string, string?> headers)
    {
        var signature = headers.GetValueOrDefault("x-twilio-email-event-webhook-signature");
        var timestamp = headers.GetValueOrDefault("x-twilio-email-event-webhook-timestamp");
        var (verified, skipped, reason) = _sendGridVerifier.Verify(rawBody, signature, timestamp);

        try { await _auditClient.IngestAsync(new IngestAuditEventRequest { EventType = "webhook.received", Action = "webhook.received", SourceSystem = "notifications", Description = "SendGrid webhook received" }); } catch { }

        if (!verified && !skipped)
        {
            _logger.LogWarning("SendGrid webhook rejected: {Reason}", reason);
            return new WebhookResult { Accepted = false, RejectedReason = "signature_verification_failed" };
        }

        var rawLog = await _webhookLogRepo.CreateAsync(new ProviderWebhookLog
        {
            Provider = "sendgrid", Channel = "email",
            RequestHeadersJson = JsonSerializer.Serialize(headers),
            PayloadJson = rawBody, SignatureVerified = verified,
            ProcessingStatus = "received", ReceivedAt = DateTime.UtcNow
        });

        var events = SendGridNormalizer.ParseEvents(rawBody);
        foreach (var rawEvent in events)
        {
            try { await ProcessSendGridEventAsync(rawEvent); } catch (Exception ex) { _logger.LogError(ex, "Error processing SendGrid event"); }
        }

        await _webhookLogRepo.UpdateStatusAsync(rawLog.Id, "processed");
        return new WebhookResult { Accepted = true };
    }

    public async Task<WebhookResult> HandleTwilioAsync(string rawBody, Dictionary<string, string?> headers, string requestUrl, Dictionary<string, string> formParams)
    {
        var signature = headers.GetValueOrDefault("x-twilio-signature");
        var (verified, skipped, reason) = _twilioVerifier.Verify(requestUrl, formParams, signature);

        try { await _auditClient.IngestAsync(new IngestAuditEventRequest { EventType = "webhook.received", Action = "webhook.received", SourceSystem = "notifications", Description = "Twilio webhook received" }); } catch { }

        if (!verified && !skipped)
        {
            _logger.LogWarning("Twilio webhook rejected: {Reason}", reason);
            return new WebhookResult { Accepted = false, RejectedReason = "signature_verification_failed" };
        }

        var rawLog = await _webhookLogRepo.CreateAsync(new ProviderWebhookLog
        {
            Provider = "twilio", Channel = "sms",
            RequestHeadersJson = JsonSerializer.Serialize(headers),
            PayloadJson = JsonSerializer.Serialize(formParams), SignatureVerified = verified,
            ProcessingStatus = "received", ReceivedAt = DateTime.UtcNow
        });

        try { await ProcessTwilioEventAsync(formParams); } catch (Exception ex) { _logger.LogError(ex, "Error processing Twilio event"); }

        await _webhookLogRepo.UpdateStatusAsync(rawLog.Id, "processed");
        return new WebhookResult { Accepted = true };
    }

    private async Task ProcessSendGridEventAsync(SendGridEventItem rawEvent)
    {
        var normalized = SendGridNormalizer.Normalize(rawEvent);
        var dedupKey = normalized.ProviderMessageId != null ? $"sendgrid:{normalized.ProviderMessageId}:{normalized.RawEventType}:{normalized.EventTimestamp.Ticks}" : null;

        if (dedupKey != null && await _eventRepo.FindByDedupKeyAsync(dedupKey) != null) return;

        NotificationAttempt? attempt = null;
        if (normalized.ProviderMessageId != null)
            attempt = await _attemptRepo.FindByProviderMessageIdAsync(normalized.ProviderMessageId);

        var notificationId = attempt?.NotificationId;
        var tenantId = attempt?.TenantId;

        await _eventRepo.CreateAsync(new NotificationEvent
        {
            TenantId = tenantId, NotificationId = notificationId, NotificationAttemptId = attempt?.Id,
            Provider = "sendgrid", Channel = "email", RawEventType = normalized.RawEventType,
            NormalizedEventType = normalized.NormalizedEventType, EventTimestamp = normalized.EventTimestamp,
            ProviderMessageId = normalized.ProviderMessageId, DedupKey = dedupKey
        });

        if (attempt != null)
        {
            await _deliveryStatusSvc.UpdateAttemptFromEventAsync(attempt.Id, normalized.NormalizedEventType);
            if (notificationId.HasValue)
                await _deliveryStatusSvc.UpdateNotificationFromEventAsync(notificationId.Value, normalized.NormalizedEventType);
        }

        if (tenantId.HasValue && normalized.RecipientEmail != null)
            await _contactHealthSvc.ProcessEventAsync(tenantId.Value, "email", normalized.RecipientEmail, normalized.NormalizedEventType, normalized.RawEventType);

        if (tenantId.HasValue && notificationId.HasValue && DeliveryFinalEvents.Contains(normalized.NormalizedEventType))
            await _deliveryIssueSvc.ProcessEventAsync(new DeliveryIssueContext { TenantId = tenantId.Value, NotificationId = notificationId.Value, NotificationAttemptId = attempt?.Id, Channel = "email", Provider = "sendgrid", NormalizedEventType = normalized.NormalizedEventType, RawEventType = normalized.RawEventType, RecipientContact = normalized.RecipientEmail });

        if (tenantId.HasValue && normalized.RecipientEmail != null && SgSuppressionMap.TryGetValue(normalized.NormalizedEventType, out var suppressionType))
        {
            try
            {
                await _suppressionRepo.UpsertFromEventAsync(new ContactSuppression
                {
                    TenantId = tenantId.Value, Channel = "email", ContactValue = normalized.RecipientEmail.Trim().ToLowerInvariant(),
                    SuppressionType = suppressionType, Reason = $"Auto-suppressed via SendGrid webhook: {normalized.RawEventType}",
                    Source = "provider_webhook", Notes = $"providerMessageId: {normalized.ProviderMessageId ?? "unknown"}"
                });
            }
            catch (Exception ex) { _logger.LogError(ex, "Failed to upsert suppression from SendGrid event"); }
        }
    }

    private async Task ProcessTwilioEventAsync(Dictionary<string, string> formParams)
    {
        // ── LS-NOTIF-SMS-003: Inbound SMS processing with tenant resolution ──────
        // Twilio sends inbound messages to the same webhook URL with Direction=inbound or inbound-api.
        // These are NOT outbound status callbacks — detect, resolve tenant, then handle separately.
        var direction = formParams.GetValueOrDefault("Direction", "");
        if (InboundDirections.Contains(direction))
        {
            var fromPhone  = formParams.GetValueOrDefault("From", "");
            var toPhone    = formParams.GetValueOrDefault("To", "");
            var body       = formParams.GetValueOrDefault("Body", "");
            var messageSid = formParams.GetValueOrDefault("MessageSid") ?? formParams.GetValueOrDefault("SmsSid");

            _logger.LogInformation("Twilio inbound SMS from {From} to {To}: SID={Sid}",
                MaskPhone(fromPhone), MaskPhone(toPhone), messageSid);

            var keyword = _smsPreferenceSvc.ClassifyKeyword(body);

            // Resolve inbound `To` number to tenant/provider config.
            // This must happen BEFORE any preference mutation to ensure correct tenant scoping.
            InboundSmsResolutionResult resolution;
            try
            {
                resolution = await _inboundSmsResolver.ResolveAsync(toPhone);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "InboundSmsResolver threw unexpectedly for To={To}", MaskPhone(toPhone));
                resolution = InboundSmsResolutionResult.Unresolved(toPhone);
            }

            if (keyword != null)
            {
                if (resolution.Resolved)
                {
                    // Resolved tenant — process keyword with full provider context.
                    try
                    {
                        await _smsPreferenceSvc.ProcessInboundKeywordWithContextAsync(new InboundSmsKeywordContext
                        {
                            TenantId          = resolution.TenantId,
                            FromPhone         = fromPhone,
                            ToPhone           = toPhone,
                            Keyword           = keyword,
                            RawKeyword        = body.Trim(),
                            ProviderMessageId = messageSid,
                            ProviderConfigId  = resolution.ProviderConfigId,
                            Provider          = resolution.Provider ?? "twilio",
                            TenantResolved    = true,
                        });
                    }
                    catch (Exception ex)
                    {
                        _logger.LogError(ex, "Failed to process inbound SMS keyword for TenantId={Tid}", resolution.TenantId);
                    }
                }
                else
                {
                    // Unresolved — do NOT mutate any tenant-scoped preference state.
                    // Emit unresolved audit event and structured log only.
                    try
                    {
                        await _smsPreferenceSvc.AuditUnresolvedInboundAsync(fromPhone, toPhone, keyword, body.Trim(), messageSid);
                    }
                    catch (Exception ex)
                    {
                        _logger.LogError(ex, "Failed to audit unresolved inbound SMS from {From}", MaskPhone(fromPhone));
                    }
                }
            }
            else
            {
                _logger.LogDebug("Twilio inbound SMS body is not a compliance keyword — ignoring. SID={Sid}", messageSid);
            }

            // Inbound messages are not outbound status events — return without further processing.
            return;
        }

        // ── Outbound status callback (existing behavior) ──────────────────────────
        var normalized = TwilioNormalizer.Normalize(formParams);
        var dedupKey = normalized.ProviderMessageId != null ? $"twilio:{normalized.ProviderMessageId}:{normalized.RawEventType}" : null;

        if (dedupKey != null && await _eventRepo.FindByDedupKeyAsync(dedupKey) != null) return;

        NotificationAttempt? attempt = null;
        if (normalized.ProviderMessageId != null)
            attempt = await _attemptRepo.FindByProviderMessageIdAsync(normalized.ProviderMessageId);

        var notificationId = attempt?.NotificationId;
        var tenantId = attempt?.TenantId;

        await _eventRepo.CreateAsync(new NotificationEvent
        {
            TenantId = tenantId, NotificationId = notificationId, NotificationAttemptId = attempt?.Id,
            Provider = "twilio", Channel = "sms", RawEventType = normalized.RawEventType,
            NormalizedEventType = normalized.NormalizedEventType, EventTimestamp = normalized.EventTimestamp,
            ProviderMessageId = normalized.ProviderMessageId, DedupKey = dedupKey
        });

        if (attempt != null)
        {
            await _deliveryStatusSvc.UpdateAttemptFromEventAsync(attempt.Id, normalized.NormalizedEventType);
            if (notificationId.HasValue)
                await _deliveryStatusSvc.UpdateNotificationFromEventAsync(notificationId.Value, normalized.NormalizedEventType);
        }

        if (tenantId.HasValue && normalized.RecipientPhone != null)
            await _contactHealthSvc.ProcessEventAsync(tenantId.Value, "sms", normalized.RecipientPhone, normalized.NormalizedEventType, normalized.RawEventType);

        if (tenantId.HasValue && notificationId.HasValue && DeliveryFinalEvents.Contains(normalized.NormalizedEventType))
            await _deliveryIssueSvc.ProcessEventAsync(new DeliveryIssueContext { TenantId = tenantId.Value, NotificationId = notificationId.Value, NotificationAttemptId = attempt?.Id, Channel = "sms", Provider = "twilio", NormalizedEventType = normalized.NormalizedEventType, RawEventType = normalized.RawEventType, RecipientContact = normalized.RecipientPhone, ErrorCode = normalized.ErrorCode, ErrorMessage = normalized.ErrorMessage });

        if (tenantId.HasValue && normalized.RecipientPhone != null && TwilioSuppressionMap.TryGetValue(normalized.NormalizedEventType, out var suppressionType))
        {
            try
            {
                var normalizedPhone = System.Text.RegularExpressions.Regex.Replace(normalized.RecipientPhone.Trim(), @"[^\d+]", "");
                await _suppressionRepo.UpsertFromEventAsync(new ContactSuppression
                {
                    TenantId = tenantId.Value, Channel = "sms", ContactValue = normalizedPhone,
                    SuppressionType = suppressionType, Reason = $"Auto-suppressed via Twilio webhook: {normalized.RawEventType}",
                    Source = "provider_webhook", Notes = $"providerMessageId: {normalized.ProviderMessageId ?? "unknown"}"
                });
            }
            catch (Exception ex) { _logger.LogError(ex, "Failed to upsert suppression from Twilio event"); }
        }
    }

    private static string MaskPhone(string phone)
    {
        if (string.IsNullOrWhiteSpace(phone)) return "***";
        var digits = System.Text.RegularExpressions.Regex.Replace(phone.Trim(), @"[^\d+]", "");
        return digits.Length > 3 ? digits[..3] + "***" : "***";
    }
}

File: _archived/notifications-nodejs/src/services/webhook-ingestion.service.ts
import { ProviderWebhookLogRepository } from "../repositories/provider-webhook-log.repository";
import { NotificationEventRepository } from "../repositories/notification-event.repository";
import { NotificationAttemptRepository } from "../repositories/notification-attempt.repository";
import { DeliveryStatusService } from "./delivery-status.service";
import { RecipientContactHealthService } from "./recipient-contact-health.service";
import { DeliveryIssueService } from "./delivery-issue.service";
import { auditClient } from "../integrations/audit/audit.client";
import { ContactSuppressionRepository } from "../repositories/contact-suppression.repository";
import { normalizeContactValue } from "../shared/contact-normalizer";
import { SuppressionType } from "../types";
import { logger } from "../shared/logger";
import { NormalizedEventType } from "../models/notification-event.model";
import { NotificationAttempt } from "../models/notification-attempt.model";
import { NotificationChannel } from "../types";
import { AppConfig } from "../config";
import {
  SendGridWebhookVerifier,
} from "../integrations/webhooks/verifiers/sendgrid.verifier";
import {
  TwilioWebhookVerifier,
} from "../integrations/webhooks/verifiers/twilio.verifier";
import {
  normalizeSendGridEvent,
  SendGridEventItem,
} from "../integrations/webhooks/normalizers/sendgrid.normalizer";
import {
  normalizeTwilioEvent,
  TwilioStatusPayload,
} from "../integrations/webhooks/normalizers/twilio.normalizer";

const DELIVERY_FINAL_EVENTS: NormalizedEventType[] = [
  "delivered",
  "bounced",
  "failed",
  "undeliverable",
  "rejected",
  "complained",
  "unsubscribed",
];

export class WebhookIngestionService {
  private webhookLogRepo: ProviderWebhookLogRepository;
  private eventRepo: NotificationEventRepository;
  private attemptRepo: NotificationAttemptRepository;
  private deliveryStatusSvc: DeliveryStatusService;
  private contactHealthSvc: RecipientContactHealthService;
  private deliveryIssueSvc: DeliveryIssueService;
  private suppressionRepo: ContactSuppressionRepository;
  private sendgridVerifier: SendGridWebhookVerifier;
  private twilioVerifier: TwilioWebhookVerifier;
  private config: AppConfig;

  constructor(config: AppConfig) {
    this.config = config;
    this.webhookLogRepo = new ProviderWebhookLogRepository();
    this.eventRepo = new NotificationEventRepository();
    this.attemptRepo = new NotificationAttemptRepository();
    this.deliveryStatusSvc = new DeliveryStatusService();
    this.contactHealthSvc = new RecipientContactHealthService();
    this.deliveryIssueSvc = new DeliveryIssueService();
    this.suppressionRepo = new ContactSuppressionRepository();

    this.sendgridVerifier = new SendGridWebhookVerifier({
      enabled: config.sendgrid.webhookVerificationEnabled,
      publicKey: config.sendgrid.webhookPublicKey,
      nodeEnv: config.nodeEnv,
    });

    this.twilioVerifier = new TwilioWebhookVerifier({
      enabled: config.twilio.webhookVerificationEnabled,
      authToken: config.twilio.webhookAuthToken || config.twilio.authToken,
      nodeEnv: config.nodeEnv,
    });
  }

  async handleSendGrid(params: {
    rawBody: string;
    headers: Record<string, string | undefined>;
    events: unknown;
  }): Promise<{ accepted: boolean; rejectedReason?: string }> {
    const { rawBody, headers } = params;

    const signature = headers["x-twilio-email-event-webhook-signature"];
    const timestamp = headers["x-twilio-email-event-webhook-timestamp"];

    const verificationResult = this.sendgridVerifier.verify(rawBody, signature, timestamp);

    const signatureVerified = verificationResult.verified;

    await auditClient.publishEvent({
      eventType: "webhook.received",
      provider: "sendgrid",
      metadata: { signatureVerified, skipped: verificationResult.skipped },
    });

    if (!verificationResult.verified && !verificationResult.skipped) {
      logger.warn("SendGrid webhook rejected — signature verification failed", {
        reason: verificationResult.reason,
      });

      await auditClient.publishEvent({
        eventType: "webhook.signature_failed",
        provider: "sendgrid",
        metadata: { reason: verificationResult.reason },
      });

      return { accepted: false, rejectedReason: "signature_verification_failed" };
    }

    const rawLog = await this.webhookLogRepo.create({
      provider: "sendgrid",
      channel: "email",
      requestHeadersJson: JSON.stringify(headers),
      payloadJson: rawBody,
      signatureVerified,
      processingStatus: "received",
      receivedAt: new Date(),
    });

    let events: SendGridEventItem[] = [];
    try {
      const parsed = JSON.parse(rawBody);
      events = Array.isArray(parsed) ? (parsed as SendGridEventItem[]) : [parsed as SendGridEventItem];
    } catch (err) {
      logger.error("Failed to parse SendGrid webhook body", { error: String(err) });
      await this.webhookLogRepo.updateStatus(rawLog.id, "failed", "JSON parse error");
      return { accepted: true };
    }

    for (const rawEvent of events) {
      await this.processSendGridEvent(rawEvent).catch((err) => {
        logger.error("Error processing SendGrid event", {
          error: String(err),
          rawEvent,
        });
      });
    }

    await this.webhookLogRepo.updateStatus(rawLog.id, "processed");
    return { accepted: true };
  }

  async handleTwilio(params: {
    rawBody: string;
    headers: Record<string, string | undefined>;
    requestUrl: string;
    formParams: Record<string, string>;
  }): Promise<{ accepted: boolean; rejectedReason?: string }> {
    const { rawBody, headers, requestUrl, formParams } = params;

    const signature = headers["x-twilio-signature"];

    const verificationResult = this.twilioVerifier.verify(requestUrl, formParams, signature);

    const signatureVerified = verificationResult.verified;

    await auditClient.publishEvent({
      eventType: "webhook.received",
      provider: "twilio",
      metadata: { signatureVerified, skipped: verificationResult.skipped },
    });

    if (!verificationResult.verified && !verificationResult.skipped) {
      logger.warn("Twilio webhook rejected — signature verification failed", {
        reason: verificationResult.reason,
      });

      await auditClient.publishEvent({
        eventType: "webhook.signature_failed",
        provider: "twilio",
        metadata: { reason: verificationResult.reason },
      });

      return { accepted: false, rejectedReason: "signature_verification_failed" };
    }

    const rawLog = await this.webhookLogRepo.create({
      provider: "twilio",
      channel: "sms",
      requestHeadersJson: JSON.stringify(headers),
      payloadJson: JSON.stringify(formParams),
      signatureVerified,
      processingStatus: "received",
      receivedAt: new Date(),
    });

    await this.processTwilioEvent(formParams as TwilioStatusPayload).catch((err) => {
      logger.error("Error processing Twilio event", { error: String(err) });
    });

    await this.webhookLogRepo.updateStatus(rawLog.id, "processed");
    return { accepted: true };
  }

  private async processSendGridEvent(rawEvent: SendGridEventItem): Promise<void> {
    const normalized = normalizeSendGridEvent(rawEvent);

    const dedupKey = normalized.providerMessageId
      ? `sendgrid:${normalized.providerMessageId}:${normalized.rawEventType}:${normalized.eventTimestamp.getTime()}`
      : null;

    if (dedupKey) {
      const existing = await this.eventRepo.findByDedupKey(dedupKey);
      if (existing) {
        logger.debug("Skipping duplicate SendGrid event", { dedupKey });
        return;
      }
    }

    // Correlate to attempt
    let attempt: NotificationAttempt | null = null;
    if (normalized.providerMessageId) {
      attempt = await NotificationAttempt.findOne({
        where: { providerMessageId: normalized.providerMessageId },
      });
    }

    const notificationId = attempt ? (attempt as any).notificationId : null;
    const tenantId = attempt ? (attempt as any).tenantId : null;

    // Persist normalized event
    await this.eventRepo.create({
      tenantId,
      notificationId,
      notificationAttemptId: attempt?.id ?? null,
      provider: "sendgrid",
      channel: "email",
      rawEventType: normalized.rawEventType,
      normalizedEventType: normalized.normalizedEventType,
      eventTimestamp: normalized.eventTimestamp,
      providerMessageId: normalized.providerMessageId,
      metadataJson: JSON.stringify(normalized.metadata),
      dedupKey,
    });

    // Status updates
    if (attempt) {
      await this.deliveryStatusSvc.updateAttemptFromEvent(attempt.id, normalized.normalizedEventType);
      if (notificationId) {
        await this.deliveryStatusSvc.updateNotificationFromEvent(notificationId, normalized.normalizedEventType);
      }
    }

    // Audit events
    if (DELIVERY_FINAL_EVENTS.includes(normalized.normalizedEventType)) {
      if (normalized.normalizedEventType === "delivered") {
        await auditClient.publishEvent({
          eventType: "delivery.confirmed",
          tenantId: tenantId ?? undefined,
          channel: "email",
          provider: "sendgrid",
          metadata: { notificationId, providerMessageId: normalized.providerMessageId ?? undefined },
        });
      } else {
        await auditClient.publishEvent({
          eventType: "delivery.permanent_failure_recorded",
          tenantId: tenantId ?? undefined,
          channel: "email",
          provider: "sendgrid",
          metadata: {
            notificationId,
            normalizedEventType: normalized.normalizedEventType,
            providerMessageId: normalized.providerMessageId ?? undefined,
          },
        });
      }
    }

    if (normalized.normalizedEventType === "unsubscribed") {
      await auditClient.publishEvent({
        eventType: "recipient.unsubscribed",
        tenantId: tenantId ?? undefined,
        channel: "email",
        provider: "sendgrid",
        metadata: { email: normalized.recipientEmail ?? undefined },
      });
    }

    // Contact health + delivery issues
    if (tenantId && normalized.recipientEmail) {
      await this.contactHealthSvc.processEvent(
        tenantId,
        "email",
        normalized.recipientEmail,
        normalized.normalizedEventType,
        normalized.rawEventType
      );
    }

    if (tenantId && notificationId && DELIVERY_FINAL_EVENTS.includes(normalized.normalizedEventType)) {
      await this.deliveryIssueSvc.processEvent({
        tenantId,
        notificationId,
        notificationAttemptId: attempt?.id ?? null,
        channel: "email",
        provider: "sendgrid",
        normalizedEventType: normalized.normalizedEventType,
        rawEventType: normalized.rawEventType,
        recipientContact: normalized.recipientEmail,
      });
    }

    // NOTIF-007: Auto-suppression from compliance events
    const SG_SUPPRESSION_MAP: Partial<Record<string, SuppressionType>> = {
      bounced: "bounce",
      complained: "complaint",
      unsubscribed: "unsubscribe",
    };
    const sgSuppressionType = SG_SUPPRESSION_MAP[normalized.normalizedEventType];
    if (tenantId && normalized.recipientEmail && sgSuppressionType) {
      const normalizedEmail = normalizeContactValue("email", normalized.recipientEmail);
      await this.suppressionRepo.upsertFromEvent({
        tenantId,
        channel: "email",
        contactValue: normalizedEmail,
        suppressionType: sgSuppressionType,
        reason: `Auto-suppressed via SendGrid webhook: ${normalized.rawEventType}`,
        source: "provider_webhook",
        expiresAt: null,
        createdBy: null,
        notes: `providerMessageId: ${normalized.providerMessageId ?? "unknown"}`,
      }).catch((err: unknown) => {
        logger.error("Failed to upsert suppression from SendGrid event", { error: String(err), normalizedEmail, sgSuppressionType });
      });
    }
  }

  private async processTwilioEvent(rawEvent: TwilioStatusPayload): Promise<void> {
    const normalized = normalizeTwilioEvent(rawEvent);

    const dedupKey = normalized.providerMessageId
      ? `twilio:${normalized.providerMessageId}:${normalized.rawEventType}`
      : null;

    if (dedupKey) {
      const existing = await this.eventRepo.findByDedupKey(dedupKey);
      if (existing) {
        logger.debug("Skipping duplicate Twilio event", { dedupKey });
        return;
      }
    }

    // Correlate to attempt by MessageSid
    let attempt: NotificationAttempt | null = null;
    if (normalized.providerMessageId) {
      attempt = await NotificationAttempt.findOne({
        where: { providerMessageId: normalized.providerMessageId },
      });
    }

    const notificationId = attempt ? (attempt as any).notificationId : null;
    const tenantId = attempt ? (attempt as any).tenantId : null;

    // Persist normalized event
    await this.eventRepo.create({
      tenantId,
      notificationId,
      notificationAttemptId: attempt?.id ?? null,
      provider: "twilio",
      channel: "sms",
      rawEventType: normalized.rawEventType,
      normalizedEventType: normalized.normalizedEventType,
      eventTimestamp: normalized.eventTimestamp,
      providerMessageId: normalized.providerMessageId,
      metadataJson: JSON.stringify(normalized.metadata),
      dedupKey,
    });

    // Status updates
    if (attempt) {
      await this.deliveryStatusSvc.updateAttemptFromEvent(attempt.id, normalized.normalizedEventType);
      if (notificationId) {
        await this.deliveryStatusSvc.updateNotificationFromEvent(notificationId, normalized.normalizedEventType);
      }
    }

    // Audit events
    if (DELIVERY_FINAL_EVENTS.includes(normalized.normalizedEventType)) {
      if (normalized.normalizedEventType === "delivered") {
        await auditClient.publishEvent({
          eventType: "delivery.confirmed",
          tenantId: tenantId ?? undefined,
          channel: "sms",
          provider: "twilio",
          metadata: { notificationId, providerMessageId: normalized.providerMessageId ?? undefined },
        });
      } else {
        await auditClient.publishEvent({
          eventType: "delivery.permanent_failure_recorded",
          tenantId: tenantId ?? undefined,
          channel: "sms",
          provider: "twilio",
          metadata: {
            notificationId,
            normalizedEventType: normalized.normalizedEventType,
            providerMessageId: normalized.providerMessageId ?? undefined,
          },
        });
      }
    }

    // Contact health + delivery issues
    if (tenantId && normalized.recipientPhone) {
      await this.contactHealthSvc.processEvent(
        tenantId,
        "sms",
        normalized.recipientPhone,
        normalized.normalizedEventType,
        normalized.rawEventType
      );
    }

    if (tenantId && notificationId && DELIVERY_FINAL_EVENTS.includes(normalized.normalizedEventType)) {
      await this.deliveryIssueSvc.processEvent({
        tenantId,
        notificationId,
        notificationAttemptId: attempt?.id ?? null,
        channel: "sms",
        provider: "twilio",
        normalizedEventType: normalized.normalizedEventType,
        rawEventType: normalized.rawEventType,
        recipientContact: normalized.recipientPhone,
        errorCode: normalized.errorCode,
        errorMessage: normalized.errorMessage,
      });
    }

    // NOTIF-007: Auto-suppression from compliance events
    const TWILIO_SUPPRESSION_MAP: Partial<Record<string, SuppressionType>> = {
      bounced: "bounce",
      complained: "complaint",
      unsubscribed: "unsubscribe",
      carrier_rejected: "carrier_rejection",
    };
    const twilioSuppressionType = TWILIO_SUPPRESSION_MAP[normalized.normalizedEventType];
    if (tenantId && normalized.recipientPhone && twilioSuppressionType) {
      const normalizedPhone = normalizeContactValue("sms", normalized.recipientPhone);
      await this.suppressionRepo.upsertFromEvent({
        tenantId,
        channel: "sms",
        contactValue: normalizedPhone,
        suppressionType: twilioSuppressionType,
        reason: `Auto-suppressed via Twilio webhook: ${normalized.rawEventType}`,
        source: "provider_webhook",
        expiresAt: null,
        createdBy: null,
        notes: `providerMessageId: ${normalized.providerMessageId ?? "unknown"}`,
      }).catch((err: unknown) => {
        logger.error("Failed to upsert suppression from Twilio event", { error: String(err), normalizedPhone, twilioSuppressionType });
      });
    }
  }
}
